vm.c

```diff
@@ -29,6 +29,32 @@
     pc = n;        \
     goto *labels[OPCODE.opcode]
 
+#define VM_PUSH(n)                           \
+    do {                                     \
+        sp--;                                \
+        env->temps[sp].type = INT;           \
+        env->temps[sp].value.vint = (int) n; \
+    } while (0)
+
+#define VM_POP(n)                               \
+    do {                                        \
+        n = (size_t) env->temps[sp].value.vint; \
+        sp++;                                   \
+    } while (0)
+
+#define VM_CALL(n)   \
+    do {             \
+        VM_PUSH(pc); \
+        VM_GOTO(n);  \
+    } while (0)
+
+#define VM_RET()    \
+    do {            \
+        from = pc;  \
```

Comment:
call has the info in stack. j instruction and ret can benefit from this 'from'.

What else context I have missed?